fili-core/src/main/java/com/yahoo/bard/webservice/druid/serializers/Util.java

```diff
@@ -24,24 +25,47 @@
      * @return  an Optional String of physical name
      */
     public static Optional<String> findPhysicalName(Dimension value, JsonGenerator gen) {
-        JsonStreamContext context = gen.getOutputContext();
         String apiName = value.getApiName();
         // Search for physical name
-        while (context != null) {
-            Object parent = context.getCurrentValue();
-            if (parent instanceof DruidQuery) {
-                return Optional.of(
-                        ((DruidQuery) parent)
+        return getDruidQuery(gen, druidQuery ->
+                        druidQuery
                                 .getDataSource()
                                 .getPhysicalTables()
                                 .iterator()
                                 .next()
                                 .getPhysicalColumnName(apiName)
-                );
+        );
+    }
+
+    /**
+     * JSON tree walk to determine if there is a nested query below the current json node or not.
+     *
+     * @param gen  the Json Generator to retrieve the tree to walk on.
+     *
+     * @return  a Boolean where true indicates there are more nested query below this node, false otherwise
+     */
+    public static Boolean hasInnerQuery(JsonGenerator gen) {
+        return getDruidQuery(gen, druidQuery -> druidQuery.getInnerQuery() != null).get();
+    }
+
+    /**
+     * JSON tree walk to find the druid query context of the current context,
+     * returns the current context if current context is a druid query.
+     *
+     * @param gen  the Json Generator to retrieve the tree to walk on.
+     *
+     * @return an Optional of DruidQuery if found, Optional.empty otherwise
+     */
+    public static <T> Optional<T> getDruidQuery(JsonGenerator gen, Function<DruidQuery, T> success) {
```

Comment:
Since this method doesn't return the `DruidQuery`, calling it `getDruidQuery` is a bit confusing. I think `withDruidQuery` is perhaps a bit better, since what it's doing is to call the `success` function _in the context of_ (or _with_) a `DruidQuery`.